File: 00046. Permutations/solution.js
/**
 * Problem: 46. Permutations
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Returns all permutations of an array of numbers
 *
 * @param {number[]} nums - Input array
 *
 * @returns {number[][]} - All permutations
 */
const permute = (nums) => {
  // Initialize the result array to store all permutations
  const permutations = []

  // Helper function to generate permutations recursively
  const generatePermutations = (index, arr, result) => {
    // If the current index reaches the end, push a copy of the array to result
    if (index === arr.length) {
      result.push([...arr])

      return
    }

    // Iterate through the array and swap elements to generate new permutations
    for (let i = index; i < arr.length; i++) {
      swapElements(i, index, arr) // Swap current element with the index
      generatePermutations(index + 1, arr, result) // Recurse for the next index
      swapElements(i, index, arr) // Swap back to restore original order
    }
  }

  // Swaps two elements in the array
  const swapElements = (a, b, arr) => {
    const temp = arr[a]
    arr[a] = arr[b]
    arr[b] = temp
  }

  // Start generating permutations from index 0
  generatePermutations(0, nums, permutations)

  // Return the array containing all permutations
  return permutations
}

File: 00046. Permutations/solution.go
/**
 * Problem: 46. Permutations
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

// permute generates all possible permutations of the input slice nums
func permute(nums []int) [][]int {
	// res stores all the permutations we'll generate
	var res [][]int
	
	// permutation will store the current permutation we're building
	// initialized with same length as input slice
	permutation := make([]int, len(nums))
	
	// visit keeps track of which numbers we've used in current permutation
	// using a boolean array where true means the number at that index is used
	visit := make([]bool, len(nums))
	
	// backtrack is our recursive helper function
	// index represents the current position we're filling in permutation
	var backtrack func(int)
	backtrack = func(index int) {
		// Base case: if we've filled all positions (index == len(nums))
		// we've completed one permutation
		if index == len(nums) {
			// Create a deep copy of current permutation
			// We need to copy because permutation slice will be modified
			copiedPermutation := make([]int, len(nums))
			copy(copiedPermutation, permutation)
			
			// Add the completed permutation to our results
			res = append(res, copiedPermutation)
			
			return
		}   
		
		// Try placing each number at the current index
		for i := 0; i < len(nums); i++ {
			// If this number hasn't been used in current permutation
			if visit[i] == false {
				// Mark this number as used
				visit[i] = true
				// Place this number at current index
				permutation[index] = nums[i]
				// Recursively fill the next position
				backtrack(index+1)   
				// Backtrack: mark number as unused for next iteration
				visit[i] = false
			}
		}
	}
	
	// Start the backtracking process from index 0
	backtrack(0)
	
	return res
}
